test/index.js

```diff
@@ -1589,6 +1589,33 @@ describe('def', function() {
                    '  - (Either ??? ???)\n'));
   });
 
+  it('supports values of "foreign" types', function() {
+    //  id :: a -> a
+    var id = def('id', {}, [a, a], R.identity);
+
+    eq(id({'@@type': 'my-package/Foreign'}), {'@@type': 'my-package/Foreign'});
+
+    throws(function() { id([{'@@type': 'my-package/Foo'},
+                            {'@@type': 'my-package/Bar'}]); },
+           errorEq(TypeError,
+                   '‘id’ received [{"@@type": "my-package/Foo"}, ' +
+                   '{"@@type": "my-package/Bar"}] as its first argument, ' +
```

Comment:
> Maybe we should have a specific error message for an `Array` with no shared consistent type for all values?

Good idea. It needn't be specific to `$.Array`, either. It could apply to any parameterized type.

In fact, now that I think about it this is really a TVCV (type-variable constraint violation). I'll see whether we can use `constraintViolation` here instead.